src/rascal/models/sparse_kernel_predict.hh

```diff
@@ -0,0 +1,256 @@
+
+    // find offsets alongs the sparse points direction
+    std::map<int, int> offsets{sparse_points.get_offsets()};
+    Keys_t rep_keys{prop_grad.get_keys()};
+    std::map<int, Keys_t> keys_intersect{};
+    for (const int & sp : species_intersect) {
+      keys_intersect[sp] =
+          internal::set_intersection(rep_keys, sparse_points.keys_sp.at(sp));
+    }
+
+    // attach partial gradients array to manager
+    auto && djfi{
+        *manager
+             ->template get_property<Property<double, 2, Manager_t, 1, ThreeD>>(
+                 djfi_name, true, true)};
+    // don't recompute the partial gradients if already up to date
+    if (djfi.is_updated()) {
+      return;
+    }
+
+    djfi.resize();
+    djfi.setZero();
+
+    if (species_intersect.empty()) {
+      return;
+    }
+
+    math::Vector_t weights_scaled(weights.size());
+    size_t i_row{0};
+    for (auto center : manager) {
+      const int a_sp{center.get_atom_type()};
+      // \alpha_n^{scaled} = \alpha_n * [z* (X_j \dot T_n)^{z-1}]
+      weights_scaled =
+          (weights.array() *
+           (zeta *
+            internal::pow_zeta(sparse_points.dot(a_sp, prop[center]), zeta - 1))
+               .transpose()
+               .array())
+              .matrix();
+      // \sum_n \alpha_n^{scaled} T_n
+      SparsePoints sparse_point_scaled{sparse_points.dot(a_sp, weights_scaled)};
+      const size_t n_neigh{center.pairs_with_self_pair().size()};
+      // sparse_point_scaled \dot dX_i/dr_j
+      if (do_block_by_key_dot) {
+        auto rep_grads = prop_grad.get_raw_data_view();
+        const auto & values_by_sp = sparse_point_scaled.values.at(a_sp);
+
+        for (const Key_t & key : keys_intersect.at(a_sp)) {
+          const auto & values_by_sp_key = values_by_sp.at(key);
+          auto spts = Eigen::Map<const math::Vector_t>(
+              values_by_sp_key.data(), static_cast<Eigen::Index>(inner_size));
+
+          math::Vector_t fij_block(n_neigh);
+
+          int col_st{prop_grad.get_gradient_col_by_key(key)};
+          for (int i_der{0}; i_der < ThreeD; i_der++) {
+            fij_block = (rep_grads.block(i_row, col_st + i_der * inner_size,
+                                         n_neigh, inner_size) *
+                         spts.transpose())
+                            .transpose();
+
+            int i_row_{0};
+            for (auto neigh : center.pairs_with_self_pair()) {
+              djfi[neigh](i_der) += fij_block(i_row_);
+              i_row_++;
+            }  // neigh
+          }    // i_der
+        }      // key
+        i_row += n_neigh;
+      } else {
+        for (auto neigh : center.pairs_with_self_pair()) {
+          djfi[neigh] =
+              sparse_point_scaled.dot_derivative(a_sp, prop_grad[neigh]);
+        }
+      }
+    }  // center
+  }
+
+  /**
+   * Compute the gradients of a structure w.r.t atomic positions
+   * using a sparse GPR model. Only GAP model is implemented at the moment (see
```

Comment:
I'm confused. What's the difference between a "generic" sparse GPR model and GAP? I mean, GAP is literally just sparse GPR applied to energies and forces.

Reply:
a GAP model is a very specific flavor of sparse GPR: Subset of regressor with a cosine kernel function that does not mix different central atomic types.

Reply:
No it's not. GAP doesn't specify the cosine kernel; if you look in the QUIP source you'll see several other kernel types implemented (dot-product, aka "cosine", along with squared-exponential, bond real-space, and a few other things). All define a GAP model, because the GAP just refers to (sparse) Gaussian process regression applied to potential energy surfaces. Nothing more, nothing less.

Reply:
you are right but in the case of GAP with SOAP I don't think any of the other possibilities that you mention have been used in an article, am I wrong ? So I have been using 'GAP model' as a short hand for 'sparse GPR additive energy model with SOAP and cosine kernel that does not mix different central atomic types'. Do you have a better and yet short alternative ?

Reply:
SOAP-GAP is what I've been calling it.